### ISSUE TITLE
[config]: Add ability to pass config object into `extends` property

### ISSUE THREAD

**Issue by avasuro** [eslint/eslint#14269]
**The version of ESLint you are using.**
any version, e.g. v.7.22.0

**The problem you want to solve.**
We have an internal corporate eslint config which is used in all our projects.But due to the fact that there are different developers on different projects, and their preferences are also different, we would like to leave them freedom of choice according to some rules, especially concerning indentation and quotes. It would be useful if we will be able to pass some configuration into our eslint config package, so some rules will be configured according this parameters.

**Your take on the correct solution to problem.**
I think the best solution will be if `extends` property in config will be able to receive not only path strings, but also another config objects, like that:

```javascript
const generateMyConfig from 'my-config';

module.exports = {
    extends: [
        // usual way to extend config is to provide it's package name or path:
        'some-eslint-config',
        // I suggest to support also config objects here:
        { extends: 'eslint:recommended', rules: [ /*...some rules here... */ ] },
        // Because if eslint will support config objects in `extends` it will be possible
        // to use config generators and pass parameters to it:
        // (example implementation of generateMyConfig see below):
        generateMyConfig({indent: 'tab', quotes: 'single', useReact: true})
    ]
    // ...other config props
}
```

```javascript
// `my-config` package

// example implementation of generateMyConfig:
// (pay attention how `indent`, `quotes` and `useReact` used in several rules 
// that is another reason why config generators may have sense)
function generateMyConfig({indent, quotes, useReact}) {
    return {
        extends:  'eslint:recommended',
        plugins: [
            useReact && 'react'
        ].filter(notEmpty),
        rules: {
             quotes: ['error', quotes, {avoidEscape: true}],
             indent: ['error', indent, {/* ... other parameters here... */}],
             ...(useReact ? {
                 'jsx-quotes': ['error', `prefer-${quotes}`],
                 'react/jsx-indent': ['error', indent],
             } : {})
        }
    }
}
module.exports = generateMyConfig;
```

**Are you willing to submit a pull request to implement this change?**
I doubt that I will be able to figure out how eslint works, so most probably no.

**Comment by nzakas** [comment]
Thanks for the suggestion. We aren’t making any further changes to the current config system because we are moving to a new config system (see #13481).
